File: example/java/client/Employees.java
/*
 * Copyright (C) 2018 Anders Lövgren (Nowise Systems).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

 /*
 * Employees.java
 *
 * Created: 2018-jun-14, 16:03:43
 * Author:  Anders Lövgren (Nowise Systems)
 */
package client;

import employees.generated.*;
import employees.*;
import java.util.List;

/**
 * Test driver for the employees service.
 *
 * @author Anders Lövgren (Nowise Systems)
 */
public class Employees {

    public static void main(String[] args) {
        EmployeesService service = new EmployeesService();
        EmployeesPortType proxy = service.getEmployeesPort();

        try {
            System.out.println("+++ Test simple listing of all users:");
            List<String> result = proxy.getIdentities();
            System.out.println("Employees (get->identities): " + result);
        } catch (Exception exception) {
            System.err.println(exception.getMessage());
        }

        try {
            System.out.println("+++ Lookup employees from username:");
            Employee result;

            result = proxy.lookupEmployee("olle");
            System.out.println("Employees (lookup->employee): " + Serialize.ToString(result));

            result = proxy.lookupEmployee("carl");
            System.out.println("Employees (lookup->employee): " + Serialize.ToString(result));

            result = proxy.lookupEmployee("adam");
            System.out.println("Employees (lookup->employee): " + Serialize.ToString(result));

            result = proxy.lookupEmployee("eva");
            System.out.println("Employees (lookup->employee): " + Serialize.ToString(result));

            Company company = result.getCompany();
            System.out.println("Employees (lookup->company): " + Serialize.ToString(company));

            Job job = result.getJob();
            System.out.println("Employees (lookup->job): " + Serialize.ToString(job));

            Employee boss = proxy.lookupEmployee(result.getBoss());
            System.out.println("Employees (lookup->boss): " + Serialize.ToString(boss));

            result = proxy.getBoss("eva");
            System.out.println("Employees (get->boss): " + Serialize.ToString(result));

            company = proxy.getCompany("eva");
            System.out.println("Employees (get->company): " + Serialize.ToString(company));

            job = proxy.getJob("eva");
            System.out.println("Employees (get->job): " + Serialize.ToString(job));
        } catch (Exception exception) {
            System.err.println(exception.getMessage());
        }

        try {
            System.out.println("+++ Find employee records by name:");
            List<Employee> result;

            result = proxy.findEmployee("Adam Svensson");
            for (Employee employee : result) {
                System.out.println("Employees (find->employee): " + Serialize.ToString(employee));
            }

            result = proxy.findEmployee("Eva Marklund");
            for (Employee employee : result) {
                System.out.println("Employees (find->employee): " + Serialize.ToString(employee));
            }
        } catch (Exception exception) {
            System.err.println(exception.getMessage());
        }

        try {
            System.out.println("+++ Dump all employees:");
            EmployeeList result;

            result = proxy.dumpEmployees();
        } catch (Exception exception) {
            System.err.println(exception.getMessage());
        }

        try {
            System.out.println("+++ Add, update and delete employee:");
            List<String> identities;

            identities = proxy.getIdentities();
            System.out.println("Employees (get->identities): " + identities);

            Employee employee = new Employee();
            employee.setCompany(new Company());
            employee.setJob(new Job());
            employee.setName("Anders Lövgren");
            employee.getCompany().setName("Nowise Systems");
            employee.getJob().setTitle("Programmer");
            proxy.addEmployee("anders", employee);

            identities = proxy.getIdentities();
            System.out.println("Employees (get->identities): " + identities);

            employee = proxy.lookupEmployee("anders");
            System.out.println("Employees (lookup->employee): " + Serialize.ToString(employee));

            employee.setName("Anders S Lövgren");
            employee.getCompany().setName("BMC-IT, Uppsala University");
            employee.getCompany().setAddress("Husargata 3, Uppsala");
            employee.getJob().setTitle("System Developer");

            proxy.updateEmployee("anders", employee);

            identities = proxy.getIdentities();
            System.out.println("Employees (get->identities): " + identities);

            employee = proxy.lookupEmployee("anders");
            System.out.println("Employees (lookup->employee): " + Serialize.ToString(employee));

            proxy.deleteEmployee("anders");

            identities = proxy.getIdentities();
            System.out.println("Employees (get->identities): " + identities);

            employee = proxy.lookupEmployee("anders");
            System.out.println("Employees (lookup->employee): " + Serialize.ToString(employee));
            
            boolean exists = proxy.hasEmployee("anders");
            System.out.println("Employees (has->employee): " + exists);
        } catch (Exception exception) {
            System.err.println(exception.getMessage());
        }
    }
}

File: source/UUP/WebService/Example/Storage/Employees.php
<?php

/*
 * Copyright (C) 2018 Anders Lövgren (Nowise Systems).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace UUP\WebService\Example\Storage;

use RuntimeException;
use UUP\WebService\Example\Types\Employee;

/**
 * The employees store.
 * 
 * This class is a simple dropin replacement for an database where we
 * normally choose to store employee records.
 *
 * @author Anders Lövgren (Nowise Systems)
 */
class Employees
{

        /**
         * The employees list.
         * @var Employee[] 
         */
        private $_employees = array();

        /**
         * Constructor.
         * @param array $employees The employees array.
         */
        public function __construct($employees)
        {
                foreach ($employees as $user => $data) {
                        $this->_employees[$user] = new Employee($data);
                }
        }

        /**
         * Get all employees.
         * @return Employees[]
         */
        public function getAll()
        {
                return $this->_employees;
        }

        /**
         * Get all identities (username).
         * @return array
         */
        public function getIdentities()
        {
                return array_keys($this->_employees);
        }

        /**
         * Get employee object.
         * @param string $user The user identity.
         * @return Employee
         */
        public function lookup($user)
        {
                return $this->_employees[$user];
        }

        /**
         * Get employee object.
         * @param string $name The person name.
         * @return Employee[]
         */
        public function find($name)
        {
                $employees = array();

                foreach ($this->_employees as $employee) {
                        if ($employee->name == $name) {
                                $employees[] = $employee;
                        }
                }

                return $employees;
        }

        /**
         * Check if user identity exists.
         * @param string $user The user identity.
         * @return boolean
         */
        public function exist($user)
        {
                return isset($this->_employees[$user]);
        }

        /**
         * Add employee record.
         * 
         * @param string $user The user identity.
         * @param Employee $employee The employee object.
         * @throws RuntimeException
         */
        public function add($user, $employee)
        {
                $this->_employees[$user] = $employee;
        }

        /**
         * Update employee record.
         * 
         * @param string $user The user identity.
         * @param Employee $employee The employee object.
         * @throws RuntimeException
         */
        public function update($user, $employee)
        {
                $this->_employees[$user] = $employee;
        }

        /**
         * Delete employee record.
         * 
         * @param string $user The user identity.
         * @throws RuntimeException
         */
        public function delete($user)
        {
                unset($this->_employees[$user]);
        }

}
